Allow settings_target across compatibility method

adnan-ali1 opened the pull request:
Changelog: Feature: Allow access to `settings_target` in compatibility method.
Docs: Omit

Closes #14527


In the info.py module, following the guidance provided in the comments for the settings_target property, this commit focuses on refining the handling of settings_target within various package methods.

When populating the settings_target property via the package_id() method, the adjustment guarantees that it remains accessible to both the info and original_info methods. By extending its accessibility to these methods, it allows for seamless utilization within the compatibility() method as well.


- [X] Refer to the issue that supports this Pull Request.
- [X] If the issue has missing info, explain the purpose/use case/pain/need that covers this Pull Request.
- [X] I've read the [Contributing guide](https://github.com/conan-io/conan/blob/develop/.github/CONTRIBUTING.md).
- [x] I've followed the PEP8 style guides for Python code.
- [ ] I've opened another PR in the Conan docs repo to the ``develop`` branch, documenting this one.

--- 
Test Case Result

Before
```
E
E           ======== Computing necessary packages ========
E           tool/0.1: Compatible package ID 30b28f4b57d074979ce37758614b0bd2ca007f31 equal to the default package ID
E           app/0.1: Forced build from source
E           Requirements
E               app/0.1#0a0d67a98e79b54a848dc56809f632a4:4acd47797ab70f037765f4aa81d97e141228872f - Build
E           Build requirements
E               tool/0.1#643d7250bb4082abb8ef774ceb2e720c:30b28f4b57d074979ce37758614b0bd2ca007f31 - Missing
E
E           ======== Installing packages ========
E           ERROR: Missing binary: tool/0.1:30b28f4b57d074979ce37758614b0bd2ca007f31
E
E           tool/0.1: WARN: Can't find a 'tool/0.1' package binary '30b28f4b57d074979ce37758614b0bd2ca007f31' for the configuration:
E           [settings]
E           arch=x86_64
E           [settings_target]
E           arch=armv7
```

After
```
====================================== test session starts ===========================
platform linux -- Python 3.10.6, pytest-6.2.5, py-1.11.0, pluggy-1.2.0
rootdir: /mnt/c/Users/AAli/Desktop/workspace/conan, configfile: pytest.ini
plugins: xdist-3.3.1
collected 1 item

conans/test/integration/package_id/test_validate.py .                                                                                                                                                       [100%]

====================================== 1 passed in 1.14s ==============================
```

Commit:
Allow settings_target across compatibility method

In the info.py module, following the guidance provided in the comments
for the settings_target property, this commit focuses on refining the
handling of settings_target within various package methods.

When populating the settings_target property via the package_id()
method, the adjustment guarantees that it remains accessible to both the
info and original_info methods. By extending its accessibility to these
methods, it allows for seamless utilization within the compatibility()
method as well.

Signed-off-by: Adnan Ali <m_adnanali_1@hotmail.com>

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/conan-io/conan?pullRequest=14532) <br/>All committers have signed the CLA.

adnan-ali1 committed:
Adding testcase for compatibility check

Signed-off-by: Adnan Ali <m_adnanali_1@hotmail.com>

memsharded commented on conans/client/graph/compute_pid.py:
```diff
@@ -63,6 +63,10 @@ def compute_package_id(node, new_config):
 
     run_validate_package_id(conanfile)
 
+    if (conanfile.info.settings_target):
```
```suggestion
    if conanfile.info.settings_target:
```
Parenthesis unnecessary

memsharded commented on conans/client/graph/compatibility.py:
```diff
@@ -120,4 +121,7 @@ def _compatible_infos(conanfile, compatibles):
                 result.append(compat_info)
+                settings_target = elem.get("settings_target")
+                if settings_target:
+                    compat_info.settings_target.update_values(settings_target)
```
``compat_info`` has been created with ``.clone()``, but it means that in some occasions it can be ``None``. Calling ``update_values()`` over it will cause a crash.
I think there is some cases that need to be covered here, for example using the external ``compatibility.py`` plugin over a normal "host" package will crash, so better cover that case too with some extra tests.

adnan-ali1 replied:
Done.

memsharded reviewed (commented):
Thanks very much for the contribution @adnan-ali1 

I think this is looking good overall, but it might still need some improvement and some other extra tests, please let us know if the comments are enough or you'd like some help with that.

adnan-ali1 commented:
@memsharded  thanks for the review. I will update the PR with the proposed changes

adnan-ali1 committed:
Check if settings_target exist before settings

Prevents the crash as in case a requirements is not
build as a build_require it does not even has a settings_target
and settings some property in null object would crash.

Update test cases to check for multiple scenerios:

Signed-off-by: Adnan Ali <m_adnanali_1@hotmail.com>

adnan-ali1 committed:
Update conans/client/graph/compute_pid.py

take in the suggestion.

Co-authored-by: James <memsharded@gmail.com>

memsharded reviewed (approved):
Changes are good, and tests are clean and comprehensive, great job, thanks very much!

AbrilRBS reviewed (approved):
@adnan-ali1 thanks a lot for your contribution, it's really appreciated!

I like the simplicity of adding this feature and the tests you provided are quite comprehensive :)
